server/tserver/src/main/java/org/apache/accumulo/tserver/ScanServer.java

```diff
@@ -246,9 +246,30 @@ public ScanServer(ScanServerOpts opts, String[] args) {
         LOG.warn(
             "Tablet metadata caching less than one minute, may cause excessive scans on metadata table.");
       }
-      tabletMetadataCache =
-          Caffeine.newBuilder().expireAfterWrite(cacheExpiration, TimeUnit.MILLISECONDS)
-              .scheduler(Scheduler.systemScheduler()).recordStats().build(tabletMetadataLoader);
+
+      // Get the cache refresh percentage property
+      // Value must be less than 100% as 100 or over would effectively disable it
+      double cacheRefreshPercentage =
+          getConfiguration().getFraction(Property.SSERV_CACHED_TABLET_METADATA_REFRESH_PERCENT);
+      Preconditions.checkArgument(cacheRefreshPercentage < cacheExpiration,
+          "Tablet metadata cache refresh percentage is '%s' but must be less than 1",
+          cacheRefreshPercentage);
+
+      var builder = Caffeine.newBuilder().expireAfterWrite(cacheExpiration, TimeUnit.MILLISECONDS)
+          .scheduler(Scheduler.systemScheduler()).recordStats();
+      if (cacheRefreshPercentage > 0) {
+        // Compute the refresh time as a percentage of the expiration time
+        // Cache hits after this time, but before expiration, will trigger a background
+        // non-blocking refresh of the entry so future cache hits get an updated entry
+        // without having to block for a refresh
+        long cacheRefresh = (long) (cacheExpiration * cacheRefreshPercentage);
+        LOG.debug("Tablet metadata refresh percentage set to {}, refresh time set to {} ms",
+            cacheRefreshPercentage, cacheRefresh);
+        builder.refreshAfterWrite(cacheRefresh, TimeUnit.MILLISECONDS);
```

Comment:
What thread will run the refresh operation?  Based on comments by @ben-manes on #4544 it seems like the JVM wide thread pool may be used, which may not be best for I/O heavy refresh that scan server will do.   If that is the case may want to use a thread pool created in the scan server for refresh.

Looking at the javadocs, when an error happens in a background reload it seems like Caffeine will log this somewhere possibly using JVM logging APIs.  Not sure where that would go.  If these will not go to slf4j then we may want to have these background refresh task try/catch/log into a slf4j logger we know about.

Was unsure if this code should be `builder = builder.refreshAfterWrite(...)`.  Looked at the javadoc and it mentions `this` is returned, so do not need to do that.

Reply:
> If these will not go to slf4j then we may want to have these background refresh task try/catch/log into a slf4j logger we know about.

By default the platform logger redirects to `java.util.logging` (jul), which if not configured would go to the console. You can redirect the platform logger using `slf4j-jdk-platform-logging`, jul (with small overhead) using `jul-to-slf4j`, and console using [sysout-over-slf4j](https://projects.lidalia.org.uk/sysout-over-slf4j/quickstart.html). It is usually good to configure the redirects in an application to ensure everything goes to your preferred logger (log4j2).

Reply:
> It is usually good to configure the redirects in an application to ensure everything goes to your preferred logger (log4j2).

Thanks for the tip @ben-manes.  Need to look into the default configs for Accumulo and see if any handling is being setup for jul.

Reply:
np! I really like having the console logger setup too just as a failsafe from someone naively using `System.out` or `e.printStackTrace()`. There are linters for that, but still nice to have as a failsafe for 3rd party code.